providers/freebsd/process_freebsd_cgo.go

```diff
@@ -0,0 +1,410 @@
+	return out
+}
+
+func getProcCWD(p *process) (string, error) {
+	procstat, err := C.procstat_open_sysctl()
+
+	if procstat == nil {
+		return "", fmt.Errorf("failed to open procstat sysctl: %w", err)
+	}
+	defer C.procstat_close(procstat)
+
+	fs, err := C.procstat_getfiles(procstat, &p.kinfo, 0)
+	if fs == nil {
+		return "", fmt.Errorf("failed to get files: %w", err)
+	}
+
+	defer C.procstat_freefiles(procstat, fs)
+
+	files := getFileStats(fs)
+	for _, f := range files {
+		if f.fs_uflags == C.PS_FST_UFLAG_CDIR {
+			return C.GoString(f.fs_path), nil
+		}
+	}
+
+	return "", nil
+}
+
+type process struct {
+	pid   int
+	fs    procFS
+	kinfo C.struct_kinfo_proc
+}
+
+func timevalToDuration(tm C.struct_timeval) time.Duration {
+	return (time.Duration(tm.tv_sec) * 1000000) + (time.Duration(tm.tv_usec) * 1000)
+}
+
+func (p *process) CPUTime() (types.CPUTimes, error) {
+	procs, err := getProcInfo(C.KERN_PROC_PID, p.PID())
+	if err != nil {
+		return types.CPUTimes{}, err
+	}
+	p.kinfo = procs[0].kinfo
+
+	return types.CPUTimes{
+		User:   timevalToDuration(p.kinfo.ki_rusage.ru_utime),
+		System: timevalToDuration(p.kinfo.ki_rusage.ru_stime),
+	}, nil
+}
+
+func (p *process) Info() (types.ProcessInfo, error) {
+	procs, err := getProcInfo(C.KERN_PROC_PID, p.PID())
+	if err != nil {
+		return types.ProcessInfo{}, err
+	}
+	p.kinfo = procs[0].kinfo
+
+	cwd, err := getProcCWD(p)
+	if err != nil {
+		return types.ProcessInfo{}, err
+	}
+
+	args, err := getProcArgs(p)
+	if err != nil {
+		return types.ProcessInfo{}, err
+	}
+
+	exe, _ := getProcPathname(p)
+
+	return types.ProcessInfo{
+		Name:      C.GoString(&p.kinfo.ki_comm[0]),
+		PID:       int(p.kinfo.ki_pid),
+		PPID:      int(p.kinfo.ki_ppid),
+		CWD:       cwd,
+		Exe:       exe,
+		Args:      args,
+		StartTime: time.Unix(int64(p.kinfo.ki_start.tv_sec), int64(p.kinfo.ki_start.tv_usec)*1000),
+	}, nil
+}
+
+func (p *process) Memory() (types.MemoryInfo, error) {
+	procs, err := getProcInfo(C.KERN_PROC_PID, p.PID())
+	if err != nil {
+		return types.MemoryInfo{}, err
+	}
+	p.kinfo = procs[0].kinfo
+
+	return types.MemoryInfo{
+		Resident: uint64(p.kinfo.ki_rssize),
+		Virtual:  uint64(p.kinfo.ki_size),
+	}, nil
+}
+
+func (p *process) User() (types.UserInfo, error) {
+	procs, err := getProcInfo(C.KERN_PROC_PID, p.PID())
+	if err != nil {
+		return types.UserInfo{}, err
+	}
+
+	p.kinfo = procs[0].kinfo
+
+	return types.UserInfo{
+		UID:  strconv.FormatUint(uint64(p.kinfo.ki_ruid), 10),
+		EUID: strconv.FormatUint(uint64(p.kinfo.ki_uid), 10),
+		SUID: strconv.FormatUint(uint64(p.kinfo.ki_svuid), 10),
+		GID:  strconv.FormatUint(uint64(p.kinfo.ki_rgid), 10),
+		EGID: strconv.FormatUint(uint64(p.kinfo.ki_groups[0]), 10),
+		SGID: strconv.FormatUint(uint64(p.kinfo.ki_svgid), 10),
+	}, nil
+}
+
+// NetworkCounters reports network stats for the process.
+func (p *process) NetworkCounters() (*types.NetworkCountersInfo, error) {
+	snmpRaw, err := os.ReadFile(p.path("net/snmp"))
```

Comment:
AFAICT, in FreeBSD `/proc/#/net/{snmp,netstat}` do not exist. The tests for this fail and the manpages don't mention the `net` path. I am going to remove it to make progress.